### PR TITLE
ICU-22541 Fix MacOS 14 default timezone issue

### PR DESCRIPTION
**Pull request by FrankYFTang**
##### Checklist

- [X] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-22541
- [X] Required: The PR title must be prefixed with a JIRA Issue number. 
- [X] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [X] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [ ] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

**Comment by FrankYFTang** [comment]
@Robbos @syg

File: icu4c/source/common/putil.cpp [hunk 1]
```diff
@@ -1175,6 +1175,18 @@ uprv_tzname(int n)
         if (ret != nullptr && uprv_strcmp(TZDEFAULT, gTimeZoneBuffer) != 0) {
             int32_t tzZoneInfoTailLen = uprv_strlen(TZZONEINFOTAIL);
             const char *tzZoneInfoTailPtr = uprv_strstr(gTimeZoneBuffer, TZZONEINFOTAIL);
+            // MacOS14 has the realpath as something like
+            // /usr/share/zoneinfo.default/Australia/Melbourne
+            // which will not have "/zoneinfo/" in the path.
+            // Therefore if we fail, we fall back to read the link which is
+            // /var/db/timezone/zoneinfo/Australia/Melbourne
+            if (tzZoneInfoTailPtr == nullptr) {
+                ssize_t size = readlink(TZDEFAULT, gTimeZoneBuffer, sizeof(gTimeZoneBuffer)-1);
```
Comment (markusicu) [inline icu4c/source/common/putil.cpp:1184]: In other words, if the new code doesn't give us a ".../zoneinfo/..." string, then we try the old code from before PR #2323. Makes sense to me.

Alternatively, I suppose we could have looked for "zoneinfo.default" as an alternate "tail". No idea if that would be better or worse.

Apple people: Any concerns?

**Review by markusicu — APPROVED** [review]
lgtm

@richgillam @pedberg-icu et al. ok?

**Comment by srl295** [comment]
Sigh, if only tzdb would have agreed (I asked) to include an actual ID in the file.  Instead, we deal with randomness such as `../posixrules`

**Review by srl295 — APPROVED** [review]

File: icu4c/source/common/putil.cpp (hunk 1 above)
Comment (srl295) [reply icu4c/source/common/putil.cpp:1184]: I think the fallback seems reasonable.

**Comment by syg** [comment]
This might not be the right place to discuss: this won't fix the Ubuntu 18.04 breakage we saw with Chrome, which was due to /etc/localtime -> /usr/share/zoneinfo/America/New_York -> /usr/share/zoneinfo/posixrules. (To compare, my Linux box (Google's corp Linux distro) has /usr/share/zoneinfo/posixrules -> /usr/share/zoneinfo/America/New_York instead.)

Are you planning to fix that breakage for 74 as well?

**Comment by srl295** [comment]
> This might not be the right place to discuss: this won't fix the Ubuntu 18.04 breakage we saw with Chrome, 

it seems like it is the right place to discuss since the ticket claims to be about that specific breakage

**Comment by markusicu** [comment]
> does this fix the Ubuntu 18.04 breakage we saw with Chrome, which was due to /etc/localtime -> /usr/share/zoneinfo/America/New_York -> /usr/share/zoneinfo/posixrules?

I think not, because we should see tzZoneInfoTailPtr != nullptr.
We should be able to fix that by changing the check to
```
if (tzZoneInfoTailPtr == nullptr ||
        uprv_strcmp(tzZoneInfoTailPtr + tzZoneInfoTailLen, "posixrules") == 0) {
```

File: icu4c/source/common/putil.cpp [hunk 2]
```diff
@@ -1175,6 +1175,18 @@ uprv_tzname(int n)
         if (ret != nullptr && uprv_strcmp(TZDEFAULT, gTimeZoneBuffer) != 0) {
             int32_t tzZoneInfoTailLen = uprv_strlen(TZZONEINFOTAIL);
             const char *tzZoneInfoTailPtr = uprv_strstr(gTimeZoneBuffer, TZZONEINFOTAIL);
+            // MacOS14 has the realpath as something like
+            // /usr/share/zoneinfo.default/Australia/Melbourne
+            // which will not have "/zoneinfo/" in the path.
+            // Therefore if we fail, we fall back to read the link which is
+            // /var/db/timezone/zoneinfo/Australia/Melbourne
+            if (tzZoneInfoTailPtr == nullptr) {
```
Comment (markusicu) [inline icu4c/source/common/putil.cpp:1183]:
```suggestion
            // MacOS14 has the realpath as something like
            // /usr/share/zoneinfo.default/Australia/Melbourne
            // which will not have "/zoneinfo/" in the path.
            // Therefore if we fail, we fall back to read the link which is
            // /var/db/timezone/zoneinfo/Australia/Melbourne
            // We also fall back to reading the link if the realpath leads to something like
            // /usr/share/zoneinfo/posixrules
            if (tzZoneInfoTailPtr == nullptr ||
                    uprv_strcmp(tzZoneInfoTailPtr + tzZoneInfoTailLen, "posixrules") == 0) {
```

**Review by richgillam — DISMISSED** [review]
I think this looks okay, but I'm interested in what Debbie Goldsmith has to say in the email discussion that accompanied this PR.  From talking to Peter, it sounds possible that we're doing something weird in the way we write the time zone files to the disk, and that therefore the issue might not be with ICU so much as with our own time-zone-updating code.

This is all stuff I personally know diddly-squat about, but with Peter leaving, I guess I'll have to learn.  But I think this is probably okay, at least for now.

**Comment by markusicu** [comment]
Test failures. Please change icu4c/source/test/depstest/dependencies.txt:
- line 25: remove realpath_function
- lines 113..115: remove the realpath_function group
- line 117: add `readlink realpath` after readdir

File: icu4c/source/common/putil.cpp (hunk 2 above)
Comment (markusicu) [reply icu4c/source/common/putil.cpp:1183]: You applied this change... and then you reverted it in your force-push: https://github.com/unicode-org/icu/compare/unicode-org:7d58692..unicode-org:dad7f7c
Why?

**Comment by richgillam** [comment]
FWIW, I added commentary to the JIRA ticket explaining the underlying cause and recommended solution for this issue.

**Review by richgillam — APPROVED** [review]
I can't speak to Markus's comments, but Frank's logic here looks correct to me.  Based on my investigation of the code (see my comments in the JIRA ticket), I think that algorithm will work perfectly well as a long-term solution.

**Review by markusicu — CHANGES_REQUESTED** [review]
Please un-revert the "posixrules" change that you accepted yesterday and reverted today.

**Comment by FrankYFTang** [comment]
opps.. I commit it on github, and then forget to pull into my local branch, and then I added the dependecy stuff and force push, causing the revert. Sorry.

**Review by markusicu — APPROVED** [review]
thanks!

**Pull request merged**